src/azul/terra.py

```diff
@@ -352,8 +439,48 @@ def _repository_endpoint(self, *path: str) -> str:
         return furl(config.tdr_service_url,
                     path=('api', 'repository', 'v1', *path)).url
 
+    def _check_response(self,
+                        endpoint: str,
+                        response: urllib3.HTTPResponse
+                        ) -> MutableJSON:
+        if response.status == 200:
+            return json.loads(response.data)
+        elif response.status == 401:
+            raise self._insufficient_access(endpoint)
+        else:
+            raise RequirementError('Unexpected response from TDR API', response.status)
+
+    def snapshot_ids_by_name(self) -> Dict[str, str]:
+        """
+        List the TDR snapshots accessible to the current credentials.
+        """
+        limit = 100
+        endpoint = self._repository_endpoint('snapshots')
+        response = self._request('GET', endpoint, fields=dict(limit=str(limit)))
+        response = self._check_response(endpoint, response)
+        # FIXME Implement paging
+        #       https://github.com/DataBiosphere/azul/issues/3093
+        require(response['total'] <= limit, 'Too many snapshots', response['total'])
+        return {
+            snapshot['name']: snapshot['id']
+            for snapshot in response['items']
+        }
+
+    @classmethod
+    def with_service_account_credentials(cls) -> 'TDRClient':
+        return cls(credentials_provider=ServiceAccountCredentialsProvider())
+
+    @classmethod
+    def with_user_credentials(cls, token: OAuth2) -> 'TDRClient':
+        return cls(credentials_provider=UserCredentialsProvider(token))
+
 
-class TerraDRSClient(DRSClient, TerraClient):
+@attr.s(auto_attribs=True, kw_only=True, frozen=True)
+class TerraDRSClient(DRSClient, TDRClient):
+    credentials_provider: CredentialsProvider = ServiceAccountCredentialsProvider()
+    http_client: urllib3.PoolManager = attr.ib(init=False)
 
-    def __init__(self) -> None:
-        super().__init__(http_client=self._http_client)
+    def __attrs_post_init__(self):
+        # Bypass FrozenInstanceError on setting attribute post-init
+        # https://www.attrs.org/en/19.3.0/init.html#post-init-hook
+        object.__setattr__(self, 'http_client', self._http_client)
```

Comment:
See if you can find a less hacky way to model this. For example, http_client could be an abstract property in the super class. You should also make it a habit of documenting as to why a hack is employed. It took me a while to catch that there is an ordering issue since _http_client depends on `credentials_provider`. If attrs is getting in the way, you don't have to use it. I think you should dissolve the provider hierarchy and merge it into the TerraClient hierarchy. Abstract properties may be the way to go here. SAMClient can remain abstract since it isn't instantiated anywhere.

Reply:
I think I found a simpler and cleaner solution that we should discuss during parking lot.